### PR TITLE
[#1452] TX Resubmission-the wallet has to periodically resubmit unmined transactions

### PR DESCRIPTION
**Pull request by LukasKorba**
- Changelog updated
- The Compact block processor's state machine has been extended to check whether there are any unmined and unexpired transactions, and it attempts to resubmit such transactions every 5 minutes.
- Since Nighthawk servers are no longer functional, demo app's config has been updated to use zec.rocks server

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Block/Actions/TxResubmissionAction.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                
+                // the last time resubmission was triggered is more than 5 minutes ago so try again
+                if diff > Constants.thresholdToTrigger {
+                    // resubmission
+                    do {
+                        for transaction in transactions {
+                            let encodedTransaction = try transaction.encodedTransaction()
+                            
+                            try await transactionEncoder.submit(transaction: encodedTransaction)
+                            logger.info("TxResubmissionAction trying to resubmit transaction")
```
Comment (HonzaR) [inline Sources/ZcashLightClientKit/Block/Actions/TxResubmissionAction.swift:52]: Just a suggestion that applies to the below failed state log, too. I'd add a transaction identifier (like height, or so) to the log so you know which attempted trx was successful and which failed.

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift [hunk 2]
```diff
@@ -108,6 +108,19 @@ class TransactionSQLDAO: TransactionRepository {
         return try await execute(query) { try ZcashTransaction.Overview(row: $0) }
     }
 
+    func findForResubmission(upTo: BlockHeight) async throws -> [ZcashTransaction.Overview] {
+        let query = transactionsView
+            .order((ZcashTransaction.Overview.Column.minedHeight ?? BlockHeight.max).desc)
+            .filter(
```
Comment (HonzaR) [inline Sources/ZcashLightClientKit/DAO/TransactionDao.swift:114]: The `.filter()` filters out the result, right? Otherwise, we'd need the opposite.

**Review by HonzaR — APPROVED** [review]
utACK. I just left two comments inline. Other than that, I see some lint warnings, which we could resolve in a separate PR as they seem unrelated to these changes.

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift (hunk 2 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/DAO/TransactionDao.swift:114]: It's the opposite, filter says which are involved.

File: Example/ZcashLightClientSample/ZcashLightClientSample/DemoAppConfig.swift [hunk 3]
```diff
@@ -18,8 +18,8 @@ enum DemoAppConfig {
         let seed: [UInt8]
     }
 
-    static let host = ZcashSDK.isMainnet ? "mainnet.lightwalletd.com" : "lightwalletd.testnet.electriccoin.co"
-    static let port: Int = 9067
+    static let host = ZcashSDK.isMainnet ? "zec.rocks" : "lightwalletd.testnet.electriccoin.co"
+    static let port: Int = 443
```
Comment (daira) [inline Example/ZcashLightClientSample/ZcashLightClientSample/DemoAppConfig.swift:22]: This should probably have been a separate PR. It's fine, but please say in the PR description that it changes this as well.

File: Example/ZcashLightClientSample/ZcashLightClientSample/Send/SendViewController.swift [hunk 4]
```diff
@@ -247,6 +247,7 @@ class SendViewController: UIViewController {
                 KRProgressHUD.dismiss()
                 loggerProxy.info("transaction created: \(pendingTransaction)")
             } catch {
+                KRProgressHUD.dismiss()
                 fail(error)
                 loggerProxy.error("SEND FAILED: \(error)")
```
Comment (daira) [inline Example/ZcashLightClientSample/ZcashLightClientSample/Send/SendViewController.swift:252]:
```suggestion
                loggerProxy.error("SEND FAILED: \(error)")
                KRProgressHUD.dismiss()
                fail(error)
```
Log failures before doing other things.

File: Sources/ZcashLightClientKit/Block/Actions/TxResubmissionAction.swift (hunk 1 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Block/Actions/TxResubmissionAction.swift:52]: Added height, thx

File: Example/ZcashLightClientSample/ZcashLightClientSample/DemoAppConfig.swift (hunk 3 above)
Comment (LukasKorba) [reply Example/ZcashLightClientSample/ZcashLightClientSample/DemoAppConfig.swift:22]: Added to the desc.

File: Example/ZcashLightClientSample/ZcashLightClientSample/Send/SendViewController.swift (hunk 4 above)
Comment (LukasKorba) [reply Example/ZcashLightClientSample/ZcashLightClientSample/Send/SendViewController.swift:252]: Updated the code with one more instance a few lines above it.

**Review by HonzaR — APPROVED** [review]
utACK. The only concern I have is that the current Zashi wallet UX prompts users with a trx submission error and still keeps them on the Send.Confirmation screen, which could lead users to retry the trx confirmation. This new feature might potentially successfully resubmit the first failed one later, resulting in two transactions. We could rethink this Zashi UX when adopting the updated SDK version.

File: Sources/ZcashLightClientKit/Block/Actions/EnhanceAction.swift [hunk 5]
```diff
@@ -28,7 +28,7 @@ final class EnhanceAction {
         if lastScannedHeight >= latestBlockHeight {
             await context.update(state: .clearCache)
         } else {
-            await context.update(state: .updateChainTip)
+            await context.update(state: .txResubmission)
```
Comment (str4d) [inline Sources/ZcashLightClientKit/Block/Actions/EnhanceAction.swift:31]: Update `docs/cbp_state_machine.puml` to reflect the changes to the state machine in this PR.

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift [hunk 6]
```diff
@@ -108,6 +108,19 @@ class TransactionSQLDAO: TransactionRepository {
         return try await execute(query) { try ZcashTransaction.Overview(row: $0) }
     }
 
+    func findForResubmission(upTo: BlockHeight) async throws -> [ZcashTransaction.Overview] {
+        let query = transactionsView
+            .order((ZcashTransaction.Overview.Column.minedHeight ?? BlockHeight.max).desc)
```
Comment (str4d) [inline Sources/ZcashLightClientKit/DAO/TransactionDao.swift:113]: We immediately filter on `minedHeight == nil`, so AFAICT this ordering does nothing.
```suggestion
```

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed de9b8b5917ec226757422912b6306afcce48836c

File: Sources/ZcashLightClientKit/DAO/TransactionDao.swift (hunk 6 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/DAO/TransactionDao.swift:113]: Removed

File: Sources/ZcashLightClientKit/Block/Actions/EnhanceAction.swift (hunk 5 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Block/Actions/EnhanceAction.swift:31]: Done, both puml and png files have been updated.

**Review by str4d — APPROVED** [review]
utACK 85c7834152b1f89d1043b0a0aac97d5ade2eb982

**Pull request merged**